Worker crashes on outbound TLS errors (unsupported protocol, weak DH key) with Node.js 25 / OpenSSL 3.x

### Describe the bug

When Haraka attempts outbound delivery to a remote SMTP server using obsolete TLS settings (TLS < 1.2, weak Diffie-Hellman parameters), the TLS error propagates uncaught and crashes the entire worker. This causes a crash loop that blocks the entire outbound queue for all recipients until the offending message is manually removed or the destination IP is added to `no_tls_hosts`.

This is the same class of bug as #3516 (Null MX record crash) — uncaught outbound errors terminating the worker — but via a different code path (TLS handshake failure rather than AggregateError handling).

### Expected behavior

When delivering to a remote MX with obsolete TLS settings, Haraka should:

- Treat the TLS handshake failure as a delivery error
- Generate a clean DEFER (with optional retry) or BOUNCE
- Continue running without crashing
- Not block delivery to all other recipients

A single outbound delivery failure must not terminate the Haraka process.

### Observed behavior

- An outbound email is successfully queued
- During delivery, the remote MX advertises STARTTLS but uses an obsolete protocol or weak DH parameters
- Node.js / OpenSSL 3.x rejects the handshake (security level 2 by default)
- The error propagates uncaught to the worker
- Haraka crashes with one of these patterns:

**Pattern 1 — obsolete TLS protocol:**
`[ERROR] [outbound] Ongoing connection failed to <ip>:25 : Error: <hash>:error:0A000102:SSL routines:ssl_choose_client_version:unsupported protocol
[CRIT] [-] [core] Error: <hash>:error:0A000102:SSL routines:ssl_choose_client_version:unsupported protocol
[NOTICE] [-] [core] Shutting down`

**Pattern 2 — weak Diffie-Hellman key:**
`[ERROR] [outbound] Ongoing connection failed to <ip>:25 : Error: <hash>:error:0A00018A:SSL routines:tls_process_ske_dhe:dh key too small
[CRIT] [-] [core] Error: <hash>:error:0A00018A:SSL routines:tls_process_ske_dhe:dh key too small
[NOTICE] [-] [core] Shutting down`

- Worker crashes, the queue is reloaded on restart, the same TLS error reoccurs → crash loop
- The entire outbound queue is blocked until the destination IP is added to `no_tls_hosts` or the message is manually removed

### Steps To Reproduce

1. Configure Haraka with outbound delivery enabled and `enable_tls=true`
2. Queue an outbound email to a recipient whose MX uses TLS < 1.2 or DH parameters < 2048 bits
3. Run on Node.js 22+ with OpenSSL 3.x (default security level 2)
4. Observe Haraka loading the queue, attempting delivery, and crashing

### System Info
`Haraka | 3.1.x
Node   | v25.x
OS     | Linux <hostname> 6.x.x x86_64 x86_64 x86_64 GNU/Linux (Ubuntu 24.04, containerized in Kubernetes)
openssl | OpenSSL 3.x.x (security level 2)`

### Additional context

This issue is severe because:

- Node.js 22+ / OpenSSL 3.x is now the default in modern deployments
- Many legitimate remote MXs still use TLS 1.0/1.1 or weak DH parameters (especially small ISPs, self-hosted servers, or specific enterprise mail servers)
- Postfix-style fallback to plaintext (`smtp_tls_security_level = may`) is the industry standard behavior per RFC 3207, but Haraka has no equivalent default
- A single message to a problematic recipient crashes the entire mail service for all users
- Operators are forced to maintain a manual `no_tls_hosts` list, with reactive intervention each time a new bad MX is encountered

**Related issues:**
- #3516 — same crash class (uncaught outbound error → worker crash) via a different code path
- #1054 — historical request for configurable TLS in outbound (closed in v3.0.0 but the underlying robustness issue remains)

--- a/Changes.md
+++ b/Changes.md
@@ -4,6 +4,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/).
 
 ### Unreleased
 
+- fix(outbound): guard against error emit after listeners removed #3554
 - fix(outbound): yield before delivery attempts #3552
 - test(outbound,conn,endpoint,server,tls_socket): added tests #3552
 - deps(various): updated to latest

--- a/test/tls_socket.js
+++ b/test/tls_socket.js
@@ -96,6 +96,49 @@ test('tls_socket', async (t) => {
                 await new Promise((resolve) => server.close(resolve))
             }
         })
+
+        await t.test('second error handler does not crash when first handler removes all listeners', () => {
+            // Regression test for issue #3553
+            const originalNetConnect = net.connect
+            const originalTlsConnect = tls.connect
+            const originalTlsValid = tls_socket.tls_valid
+
+            const fakeCrypto = new EventEmitter()
+            fakeCrypto.writable = true
+            fakeCrypto.removeAllListeners = EventEmitter.prototype.removeAllListeners
+            fakeCrypto.setTimeout = () => {}
+            fakeCrypto.setKeepAlive = () => {}
+
+            let capturedCleartext
+            net.connect = () => fakeCrypto
+            tls.connect = () => {
+                capturedCleartext = new EventEmitter()
+                capturedCleartext.writable = true
+                capturedCleartext.setTimeout = () => {}
+                capturedCleartext.setKeepAlive = () => {}
+                return capturedCleartext
+            }
+            tls_socket.tls_valid = false
+
+            try {
+                const socket = tls_socket.connect({ host: 'bad-tls.example.com', port: 25 })
+
+                // Simulate what release_client does: strip all listeners on first error
+                socket.once('error', () => socket.removeAllListeners())
+
+                socket.upgrade({}, () => {})
+
+                // capturedCleartext now has two 'error' handlers (on from upgrade, once from attach).
+                // Emitting error must NOT throw even though the first handler removes all
+                // listeners from the outer socket before the second fires.
+                const tlsError = new Error('dh key too small')
+                assert.doesNotThrow(() => capturedCleartext.emit('error', tlsError))
+            } finally {
+                net.connect = originalNetConnect
+                tls.connect = originalTlsConnect
+                tls_socket.tls_valid = originalTlsValid
+            }
+        })
     })
 
     await t.test('getSocketOpts', async (t) => {

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -76,7 +76,7 @@ class pluggableStream extends stream.Stream {
         this.targetsocket.once('error', (exception) => {
             this.writable = this.targetsocket.writable
             exception.source = 'tls'
-            this.emit('error', exception)
+            if (this.listenerCount('error') > 0) this.emit('error', exception)
         })
         this.targetsocket.on('timeout', () => {
             this.emit('timeout')

